Guard non_overlapping_and_dense with support_as_strided.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49395 Guard non_overlapping_and_dense with support_as_strided.**

Differential Revision: [D25554953](https://our.internmc.facebook.com/intern/diff/D25554953)

Commit:
Guard non_overlapping_and_dense with support_as_strided.

[ghstack-poisoned]

swolchok commented:
I'm not a good reviewer -- don't know if this is a good idea or not.

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -1438,7 +1438,7 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
 
   bool is_non_overlapping_and_dense() const {
-    return is_non_overlapping_and_dense_;
+    return support_as_strided() && is_non_overlapping_and_dense_;
```
I don't understand this change.

If you're trying to enforce an invariant that `is_non_overlapping_and_dense_` is only true if `support_as_strided()`, then the test `TORCH_INTERNAL_ASSERT_DEBUG_ONLY(!(!support_as_strided() && is_non_overlapping_and_dense_))` seems better.

ailzhang replied:
Originally was trying to hide this check from call sites. But yeah it makes more sense to be an check.

ezyang replied:
No, I'm even more confused now.

In my previous comment, what I was trying to establish is an internal invariant on Tensor irrespective of the use sites, that whenever !support_as_strided(), then is_non_overlapping_dense_ is false. When this invariant holds, you DON'T need to put a test in the call site: when is_non_overlapping_and_dense is true, then you can also infer that support_as_strided() is true (one verifies the other). Can we try to establish that invariant? That will lead to a cleaner internal representation and also help @ngimel with devirtualization.

ngimel replied:
Yeah, in #49303 I propose that all TensorImpl derivative classes correctly set bitfields for `is_contiguous` and `is_non_overlapping_and_dense` so that those calls don't need to be virtual. `support_as_strided()` currently is super misleading - it returns that tensor is not XLA, but OpaqueTensorImpl and SparseTensorImpl and possibly BatchedTensorImpl also don't support as strided. 
So, to recap, I agree with @ezyang that we should try to establish invariant that all Impls that don't know how to deal with the concept of `is_contiguous` or `is_non_overlapping_and_dense` should set those bitfields to false, and then the function calls won't have to be guarded on anything.

ailzhang replied:
@ngimel @ezyang hmmm I'm fine with setting is_non_overlapping_and_dense to false on XLATensorImpl. (need to double check on is_contiguous). 
Another quick thought: this invariant seem to impose an invariant like
"if you don't support as_strided, set these bitfields to false.". And to me it seems this list of bitfields to set is growing over time, but all of these fields are actually try to represent the memory_format information. 
Alternatively is it better that we have a bitfield for support_as_strided so that every TensorImpl can set, and then is_contiguous/is_non_overlapping_and_dense can just be
```
  return support_as_strided && <check memory format>
```
What do you think?

ngimel replied:
Setting `support_as_strided` as a bitfield makes sense to me, but we don't have available bitfields left, so that would mean increasing TensorImpl size?

ezyang replied:
We have more bitfields left. Due to alignment concerns TensorImpl will always be rounded to a machine word, which means you get 64 bit fields per word.

That being said, I think we should be deliberate and holistic about what bit fields we add to TensorImpl. Every new computed bit field is another place for invariants to fail you. We need to clearly describe what these invariants are, and make it difficult for people to accidentally break them. This is not a small PR, it is going through all of these fields and methods with a fine toothed comb.

Commit:
Update on "Guard non_overlapping_and_dense with support_as_strided."


Differential Revision: [D25554953](https://our.internmc.facebook.com/intern/diff/D25554953)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -33,7 +33,7 @@ static inline Tensor to_impl(const Tensor& self, const TensorOptions& options, b
 
   if (memory_format == MemoryFormat::Preserve) {
-    if (self.is_non_overlapping_and_dense()) {
+    if (self.unsafeGetTensorImpl()->support_as_strided() && self.is_non_overlapping_and_dense()) {
```
You don't need this test now, right?

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -1438,6 +1438,7 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   }
 
   bool is_non_overlapping_and_dense() const {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(support_as_strided());
```
If you're serious about this invariant, it (1) needs docs, and (2) it is better to test for the invariant at sites which modify `is_non_overlapping_and_dense` or `support_as_strided`, which will reduce the number of redundant tests (no point in rechecking the invariant on every access to the tensor, when the only time the invariant could be violated is on writes)

ailzhang replied:
@ezyang I took another look at TensorImpl and I'm not so sure about this invariant now. All of following fields seem to be dependent on assumption that support as_strided, if we want to enforce the invariant we need to do it for all of them. (is this list growing? ) 
Plus technically we only need to set support_as_strided once, while all the following fields can change with tensors, now I'm thinking we should use support_as_strided as bitfield and let the following fields condition on it.
```
 6   inline void init_bitfields() {
  5     is_channels_last_ = false;
  4     is_channels_last_contiguous_ = false;
  3     is_channels_last_3d_ = false;
  2     is_channels_last_3d_contiguous_ = false;
  1     is_non_overlapping_and_dense_ = true;
```

ezyang replied:
Yeah, this is what I mean by "holistic thinking" here. If we're going to make changes here, all of the fields are going to be involved. It's better to work out what the correct invariants globally are, and then do a PR that gets us into this universe, as opposed to a smaller, "focused" patch that does the bare minimum to fix some problem. The latter approach will result in an endless procession of further bitfixes; if we can establish the correct invariants once and for all, we don't have to fix things.

I haven't thought carefully about how things should be organized. If you want a second opinion here, shout and I'll do the work.
